File: Tests/Review1/trie/trie.ts
class TrieNode {
  children: Map<string, TrieNode>;
  is_end_of_word: boolean;

  constructor() {
    this.children = new Map();
    this.is_end_of_word = false;
  }
}

class Trie {
  root: TrieNode;

  constructor() {
    this.root = new TrieNode();
  }

  insert(word: string) {
    let node = this.root;
    for (const char of word) {
      if (!node.children.has(char)) {
        node.children.set(char, new TrieNode());
      }
      node = node.children.get(char)!;
    }
    node.is_end_of_word = true;
  }

  search(word: string) {
    let node = this.root;
    for (const char of word) {
      if (!node.children.has(char)) {
        return false;
      }
      node = node.children.get(char)!;
    }
    return node.is_end_of_word;
  }

  startsWith(prefix: string) {
    let node = this.root;
    for (const char of prefix) {
      if (!node.children.has(char)) {
        return false;
      }
      node = node.children.get(char)!;
    }
    return true;
  }

  findLongestPrefix() {
    let prefix = '';
    let node = this.root;

    while (!node.is_end_of_word && node.children.size === 1) {
      const [current_char, current_child] = [...node.children.entries()][0];
      prefix += current_char;
      node = current_child;
    }

    return prefix;
  }
}

// test
const trie = new Trie(); // init

// insert words
const words_to_insert = ['apple', 'app', 'apricot', 'banana', 'bat', 'bath', 'cat', 'cut'];
for (const word of words_to_insert) {
  trie.insert(word);
}

// search particular word
console.log('------');
const search_words = ['apple', 'app', 'apex', 'bat', 'bath', 'cut', 'cute'];
for (const word of search_words) {
  console.log(trie.search(word));
}

// Word 'apple' found in Trie: True
// Word 'app' found in Trie: True
// Word 'apex' found in Trie: False
// Word 'bat' found in Trie: True
// Word 'bath' found in Trie: True
// Word 'cut' found in Trie: True
// Word 'cute' found in Trie: False

// check prefix
console.log('------');
const prefixes_to_check = ['app', 'ba', 'bat', 'cu', 'cy'];
for (const prefix of prefixes_to_check) {
  console.log(trie.startsWith(prefix));
}

// Prefix 'app' found in Trie: True
// Prefix 'ba' found in Trie: True
// Prefix 'bat' found in Trie: True
// Prefix 'cu' found in Trie: True
// Prefix 'cy' found in Trie: False

File: Data Structure & Algorithm/Archive/Template/Trie/Trie.go
// Trie

package main

import "fmt"

// TrieNode definition
type TrieNode struct {
	children    map[rune]*TrieNode
	isEndOfWord bool
}

// NewTrieNode creates a new TrieNode
func NewTrieNode() *TrieNode {
	return &TrieNode{
		children:    make(map[rune]*TrieNode),
		isEndOfWord: false,
	}
}

// Trie definition
type Trie struct {
	root *TrieNode
}

// NewTrie creates a new Trie
func NewTrie() *Trie {
	return &Trie{
		root: NewTrieNode(),
	}
}

// Insert inserts a word into the Trie
func (t *Trie) Insert(word string) {
	node := t.root
	for _, char := range word {
		if _, exists := node.children[char]; !exists {
			node.children[char] = NewTrieNode()
		}
		node = node.children[char]
	}
	node.isEndOfWord = true
}

// Search checks if a word is in the Trie
func (t *Trie) Search(word string) bool {
	node := t.root
	for _, char := range word {
		if _, exists := node.children[char]; !exists {
			return false
		}
		node = node.children[char]
	}
	return node.isEndOfWord
}

// StartsWith checks if there is any word in the Trie that starts with the given prefix
func (t *Trie) StartsWith(prefix string) bool {
	node := t.root
	for _, char := range prefix {
		if _, exists := node.children[char]; !exists {
			return false
		}
		node = node.children[char]
	}
	return true
}

// FindLongestCommonPrefix finds the longest common prefix in the Trie
func (t *Trie) FindLongestCommonPrefix() string {
	node := t.root
	prefix := ""
	for !node.isEndOfWord && len(node.children) == 1 {
		for char, nextNode := range node.children {
			prefix += string(char)
			node = nextNode
			break
		}
	}
	return prefix
}

// test
func main() {
	trie := NewTrie() // Initialize

	// Insert words
	wordsToInsert := []string{"apple", "app", "apricot", "banana", "bat", "bath", "cat", "cut"}
	for _, word := range wordsToInsert {
		trie.Insert(word)
	}

	// Search particular words
	searchWords := []string{"apple", "app", "apex", "bat", "bath", "cut", "cute"}
	for _, word := range searchWords {
		fmt.Printf("Word '%s' found in Trie: %v\n", word, trie.Search(word))
	}

	// Check prefix
	prefixesToCheck := []string{"app", "ba", "bat", "cu", "cy"}
	for _, prefix := range prefixesToCheck {
		fmt.Printf("Prefix '%s' found in Trie: %v\n", prefix, trie.StartsWith(prefix))
	}

	// Find the longest common prefix
	fmt.Printf("Longest common prefix: %s\n", trie.FindLongestCommonPrefix())
}